### PR TITLE
continue coherence work

### PR DESCRIPTION
**Pull request by lcnr**
still a lot of TODOs, opening for review (and for me to be able to go to bed again)

### CODE REVIEW

File: crates/formality-prove/src/prove/is_local.rs [hunk 1]
```diff
@@ -61,7 +62,7 @@ judgment_fn! {
         (
             // In principle this rule could be removed and preserve soundness,
             // but then we would accept code that is very prone to semver failures.
-            (may_not_be_provable(is_local_trait_ref(decls, env, assumptions, goal)) => ())
+            (may_not_be_provable(&env, |env| is_local_trait_ref(decls, &env, assumptions, goal)) => ())
             --- ("may be added by upstream in a minor release")
             (may_be_remote(decls, env, assumptions, goal) => ())
         )
```
Comment (lcnr) [inline crates/formality-prove/src/prove/is_local.rs]: this is "could an upstream crate (a crate we know about) add an impl in a non breaking change"

File: crates/formality-prove/src/prove/is_local.rs [hunk 2]
```diff
@@ -51,6 +51,7 @@ judgment_fn! {
         goal: TraitRef,
     ) => () {
         debug(assumptions, goal, decls, env)
+        assert(env.is_in_coherence_mode())
 
         (
             (may_be_downstream_trait_ref(decls, env, assumptions, goal) => ())
```
Comment (lcnr) [inline crates/formality-prove/src/prove/is_local.rs]: where this instead asks "could this get added by a source we don't know about"

File: src/test/coherence_overlap.rs [hunk 3]
```diff
@@ -275,3 +275,70 @@ fn T_and_Local_Bar_T() {
...
+                trait Foo<ty U> { }
+            },
+            crate foo {
+                struct LocalType {}
+                impl Foo<LocalType> for () {}
+
+                trait Overlap<ty U> {}
+                impl<ty T, ty U> Overlap<U> for T
+                where
+                    <T as Project>::Assoc: Foo<U> {}
```
Comment (lcnr) [inline src/test/coherence_overlap.rs:324]: `<?t as Project>::Assoc: Foo<LocalType>` normalizes to `(): Foo<LocalType>`, but `<?t as Project>::Assoc` cannot be normalized in coherence mode

Comment (lcnr) [reply src/test/coherence_overlap.rs:324]: need to figure out why this one still fails

**Comment by nikomatsakis** [comment]
Will look at this soon!

**Pull request merged**